notebooks/problem-sets/quantum_enigma_004.ipynb

```diff
@@ -0,0 +1,334 @@
+    "</details>\n",
+    "\n",
+    "<details>\n",
+    "    <summary>Hint 2</summary>\n",
+    "    You can use the following circuit that is the equivalent of a multicontrolled Hadamard gate.\n",
+    "</details>\n",
+    "<!-- ::: -->\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister\n",
+    "import matplotlib\n",
+    "from qiskit.circuit.library import MCXGate\n",
+    "from math import pi\n",
+    "\n",
+    "problem_qc = QuantumCircuit(3, 3)\n",
+    "\n",
+    "problem_qc.ry(pi/4, 2)\n",
+    "problem_qc.ccx(0, 1, 2)\n",
+    "problem_qc.ry(-pi/4, 2)\n",
+    "\n",
+    "problem_qc.draw(\"mpl\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister\n",
+    "import matplotlib\n",
+    "from qiskit.circuit.library import MCXGate\n",
+    "from math import pi\n",
+    "\n",
+    "problem_qc = QuantumCircuit(9, 9)\n",
+    "\n",
+    "# Start your work here.\n",
+    "# We've provided the circuit that is shown above\n",
+    "# Your circuit MUST be named problem_qc\n",
+    "\n",
+    "\n",
+    "\n",
+    "problem_qc.draw(\"mpl\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "## Problem 3\n",
+    "\n",
+    "*Time travel*\n",
+    "\n",
+    "One very important aspect of quantum computing is that all quantum logic gates have an inverse. This means that it is possible to simulate time traveling by going to the end of an algorithm and coming back at the start simply using the inverse of every gate in a backward manner. \n",
+    "\n",
+    "The following circuit shows the algorithm seen in the video with an extra qubit used for deciding which door will be opened in the case the diamond is in safe 2 at the beginning. The circuit has been writen up to the point in time a safe has been opened by Kettu."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister\n",
+    "import matplotlib\n",
+    "from qiskit.circuit.library import MCXGate\n",
+    "from math import pi\n",
+    "\n",
+    "qreg_q = QuantumRegister(5, 'q')\n",
+    "creg_c = ClassicalRegister(1, 'c')\n",
+    "creg_d = ClassicalRegister(1, 'd')\n",
+    "creg_f = ClassicalRegister(1, 'f')\n",
+    "creg_g = ClassicalRegister(1, 'g')\n",
+    "problem_qc = QuantumCircuit(qreg_q, creg_c, creg_d, creg_f, creg_g)\n",
+    "\n",
+    "problem_qc.ry(1.910633, 0)\n",
+    "problem_qc.h(4)\n",
```

Comment:
Maybe adding the following line above 186
#q4 is used to decide which door will be opened in the case the diamond is in safe 2 at the beginning. This is necessary since we don't we this information to be lost when measuring q3.